### ISSUE TITLE
[JENKINS-9688] Build runs on master without executor and get stuck

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#10951]
<p>In the ASF Jenkins, we do not allow builds to run on master. Thus, we have set the number of executors on master to zero. Even so, one build recently (<a href="https://builds.apache.org/hudson/job/Axis2/774/" class="external-link" target="_blank" rel="nofollow noopener">https://builds.apache.org/hudson/job/Axis2/774/</a>) got assigned to run on master. The build is configured with a label to run on one of the slaves, but somehow Jenkins assigned it incorrectly. <br/>
Running on master also meant it got stuck, presumably due to the lack of executors, without any way of stopping it.</p>

**Attachments:**

- [job_time_null.png](https://issues.jenkins.io/secure/attachment/21903/job_time_null.png)

---
<details><summary><i>Originally reported by <a title="protocol7b" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=protocol7b">protocol7b</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-9688" target="_blank">Build runs on master without executor and get stuck</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Stephen Connolly" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly">stephenconnolly</a>
<li><b>reported by</b>: protocol7b
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Incomplete
<li><b>resolved</b>: 2014-10-12T16:27:15+00:00
<li><b>votes</b>: 9
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=veithen">veithen</a>:</i>
<p>Some additional observations for this issue:</p>

<p>The Axis2 builds (there are in total 3 builds for 3 different branches) use locks. What we see is that sometimes two (or more) different builds (for different branches) where triggered at the same time by the completion of a common upstream build (Axiom trunk e.g.). Since the Axis2 builds use a common lock, one would expect that only one starts execution, while the others remain in the build queue. However, what happens is that sometimes, two builds start execution in parallel, with one waiting for the lock (i.e. instead of waiting in the build queue, it is assigned to an executor and waiting there).</p>

<p>Here is a screenshot that shows the problem (The three Axis2 builds all use the same lock):</p>

<p><a href="http://people.apache.org/~veithen/axis2-builds.png" class="external-link" target="_blank" rel="nofollow noopener">http://people.apache.org/~veithen/axis2-builds.png</a></p>

<p>While the build is waiting on the executor to acquire the lock, it is reported as running on the master. E.g. the axis2-1.6 #225 build showed:</p>

<p>"Started 42 min ago<br/>
Build is being executed for 42 min on master"</p>

<p>In that particular case, the axis2-1.6 build eventually completed successfully. However, I think (I'm 90% sure) that I saw another Axis2 build that was triggered after the axis2-1.6 #225, but that actually started execution before. That would mean that when blocked builds are assigned to executors, they are no longer executed in FIFO style.</p>

<p>As a conclusion, I think that in order to make progress on this issue, one should first concentrate on the issues that occur when using locks:</p>
<ul>
	<li>The fact that builds waiting for a lock (on an executor) may be reported as running on master makes it hard to debug things if something gets stuck.</li>
	<li>The fact that execution is not FIFO means that a build may appear to be stuck simply because there is a constant flow of other builds in the queue that use the same lock. (Unfortunately it is not possible to check if that is a valid explanation for the original issue reported in this JIRA)</li>
</ul>


<p>Note: these observations where made with the following build: Jenkins ver. 1.447-SNAPSHOT (private-01/01/2012 20:43 GMT-olamy)</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=veithen">veithen</a>:</i>
<p>Still occurs (with identical symptoms) on Jenkins 1.454.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wbauer">wbauer</a>:</i>
<p>I see the same issue sporadically (Jenkins 1.454 currently).<br/>
The build seems to be stuck in the assignment phase and no thread or sockets refers to it at all.<br/>
The build is not visible in any queue or listing except on the job itself.<br/>
The build shows:<br/>
   "Started 2 hr 1 min ago<br/>
   Build is being executed for null on master"<br/>
.<br/>
This build is not killable meaning that I have to restart the master to get rid of the build...</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=josesa">josesa</a>:</i>
<p>The problem still occurs in 1.458 see:<br/>
<span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/21903/21903_job_time_null.png" style="border: 0px solid black" /></span></p>

<p>We have multiple jobs using the same lock representing a test environment, but without any quiet period. These jobs are not meant to run on master, but on a build slave, but in the build status it shows:<br/>
"Started 1 day 14 hr ago<br/>
Build is being executed for null on master"</p>

<p>Maybe activating the default quiet period (30 seconds) will help in this situation.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=josesa">josesa</a>:</i>
<p>Even with delays it got stuck (maybe because I used exact same quiet period). Now I've configured different quiet periods to avoid this.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>:</i>
<p>Too old and too little detail in the issue report to help further investigation, so resolving as incomplete.</p>

<p>Please file a new issue if this happens with a recent version of Jenkins (no older than ten weeks or so), and provide relevant information: Installed plugins, job configuration, how this can be reproduced, etc. See also the wiki for what other information may be relevant:<br/>
<a href="https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue</a></p>
